### ISSUE TITLE
Run prisma migrate command programmatically

### ISSUE THREAD

**Issue by dested** [prisma/migrate#439]
## Problem

It would be nice to have a way to run the migrate programmatically. I am running into an issue where my api is using lambdas (serverless) and my database is only accessible from within an aws vpc. Because of this I cannot "execute commands" on deploy in an easy way. My current solution is to ssh into an ec2 that is within the vpc, execute the prisma commands, and then spin the ec2 back down. While this is working, it is far from ideal. If I was instead able to have a separate lambda that had all the migration data bundled with it I could execute the migration from a lambda (that is within the vpc). 

While this may seem like a specialized use case, databases that are not publicly accessible are common place, and solutions for migration that don't include running npm commands from the cli I think become critical.

## Solution
```
import {migrate} from '@prisma/cli';
migrate('up');
```

## Alternatives
Stated above, no great alternatives for common usecases.

**Comment by medelman17** [comment]
Seconded. @dested -- are you using one of the 'serverless' aurora clusters?

**Comment by dested** [comment]
@medelman17 Yes, and its been a bit of a struggle. I have had to move away from prisma migrate for the time being and use a different solution.

**Comment by kokokenada** [comment]
I'm facing the exact same issue.  I was trying to work around the issue by putting a db-deploy end point and running a shell as described here.  I'm hitting on odd snag.  Here's the function
````
import { APIGatewayProxyHandler } from 'aws-lambda';

export const handler: APIGatewayProxyHandler = async (event, context) => {
  try {
    if (event?.queryStringParameters?.secret !== process.env['DB_DEPLOY_SECRET']) {
      console.warn(`Invalid secret given ${event?.queryStringParameters?.secret}`);
      return {
        statusCode: 401,
        body: JSON.stringify({ message: `missing or invalid secret` }),
      };
    }

    console.log(`process.cwd()=${process.cwd()}`);

    process.env['PATH'] = process.env['PATH'] + ':' + process.env['LAMBDA_TASK_ROOT'];
    // https://aws.amazon.com/blogs/compute/running-executables-in-aws-lambda/

    console.log('process.env:');
    console.log(process.env);
    console.log('running ./db-deploy.sh');

    const { execSync } = require('child_process');
    console.log('ls -l');
    console.log(execSync('ls -l').toString());

    console.log('node -v');
    console.log(execSync('node -v').toString());

    console.log('npx -v');
    console.log(execSync('npx -v').toString());

    const retVal = execSync('./db-deploy.sh');
    console.log(retVal.toString());

    return {
      statusCode: 200,
      body: JSON.stringify({ message: `Success data!!!!! = ${retVal}` }),
    };
  } catch (e) {
    console.error(e);
    return {
      statusCode: 500,
      body: JSON.stringify({ message: `Fail = ${JSON.stringify(e, null, 2)}` }),
    };
  }
};

````

All the diagnostic outputs look good, but the prisma migrate command fails with Cannot find module '../package.json'

Hmmmm....

**Comment by kokokenada** [comment]
@schickling ^ FYI

**Comment by KrishnaPG** [comment]
Recently I was working on Knex Migration API. Prisma Migration can benefit heavily if it can be run programmatically completely. That is:

- The schemas should be input as strings to the migration function, and the migration should run without having to rely on the external file system folders. (i.e. the migration storage could be abstracted out so that user can plugin any storage system, similar to Knex Migration API)

This enables complete end-to-end functionality such as:

- Allow users to define schemas on a website UI (text box), and send it server to generate the database schemas on the fly, and expose them as GraphQL or REST API using the Prisma 2 

The migrations, if required, could be stored inside the database itself as `meta` tables and expose it to the user to rollback or migrate up inside Web UI. Or they could be stored or S3 or similar, based on user-configured storage systems inside their application.

**Comment by divyenduz** [comment]
Closing this in favor of https://github.com/prisma/migrate/issues/400

**Comment by flybayer** [comment]
We would definitely use this in Blitz. Currently we are spawning the prisma CLI.

**Comment by allesklarbeidir** [comment]
For everyone who's interested in a kind of hacky but working workaround, here are the steps needed to make it work:

1) Ensure that "migration-engine-rhel-openssl-1.0.x" is bundled in the bundle when uploading to AWS Lambda. Because even if you manage to instruct prisma to download the binaries on AWS Lambda, that's not possible because on Lambda you have read-only access only to the file system (except /tmp but that doesn't help).

That was the first problem.
There seems to be no way to tell prisma, that this binary should be downloaded in addition to the binary for the development-host ("native"). So, in order to include the binary in the bundle I needed to call the download function myself and run that script with the packagerOptions in serverless.yml:

```
    webpack:
        includeModules: true
        packager: yarn
        packagerOptions:
            scripts: # this is the magic
              - prisma generate 
              - node ./prisma/setup.awslambda.js # this is my script
```

  The script itself:
```js
// ./prisma/setup.awslambda.js
const path = require("path");
const { download } = require("@prisma/fetch-engine");

(async () => {
    
    process.env.PATCH_BRANCH = "master";
    await download({
        binaries: { "migration-engine": path.resolve("./node_modules/@prisma/cli/") },
        binaryTargets: ["rhel-openssl-1.0.x"],
        ignoreCache: true,
        printVersion: false,
        failSilent: false,
    });

    process.exit();

})();
```

So with this the migration-engine binary is bundled and exists when executing "prisma migrate".

2. Don't run `prisma migrate` via the symlink in node_modules/.bin/prisma but instead use `node_modules/@prisma/cli/build/index.js` to avoid running into the error @kokokenada encountered.

3. Patch `node_modules/@prisma/cli/build/index.js` with a `replace("async downloadBinaries(){", "async downloadBinaries(){return;")`. This avoids that prisma again wants to download the binaries on a read-only fs when running the cli.

4. Run your prisma migrate command via spawn and skip prompts (if you want, or handle them somehow).
My script looks like this. It's deployed to AWS Lambda as a seperate Function and called via the AWS Lambda Console to run the migrate command.

```js
import { spawn } from "child_process";
import path from "path";

export const lambdaHandler = async (event, context) => {
    
    return await new Promise((resolve, reject) => {

        const proc = spawn(path.resolve("./node_modules/@prisma/cli/build/index.js"), ["migrate", "up", "--create-db", "--auto-approve", "--experimental"], {
            env: {
                ...process.env,
            }
        });
    
        proc.stdout.on("data", (data) => {
            console.log(`data: ${data}`);
        });
        proc.stderr.on("data", (data) => {
            console.error(`stderr: ${data}`);
        });
        proc.on("close", (code) => {
            console.log(`exited with code ${code}`);
            
            resolve();
        });
    });

}

```


There were many pitfalls on the way to make this work. And I still need to add the "Patch `node_modules/@prisma/cli/build/index.js`" to the setup.awslambda.js script to make it work automatically.

I strongly support @dested suggestion, that this should be callable programmatically and what @KrishnaPG says sounds very appealing to me as well.

**Comment by eemelipa** [comment]
I am creating a new db for users that register to my app. I have tackled migrations by running a shell script from node, and the shell script runs the migrate cmd. Similar approach to what @flybayer and @allesklarbeidir already pointed out

Here are the scripts:
createDatabaseForUser.sh
```
#!/bin/bash

export DATABASE_URL="$1"
npx prisma migrate up --experimental --create-db --verbose
```
PrismaService.ts
```
export class PrismaService {

    static migrateUp(userId: string): Promise<void> {

        Log.info('Migrating user id: ' + userId);
        return new Promise((resolve, _) => {

            const childProcess = execFile(
                path.resolve(process.cwd(), './src/createDatabaseForUser.sh'),
                [process.env.DATABASE_URL + 'user-' + userId],
                (error, stdout, stderr) => {
                    Log.info('Migration up done.');
                    if (error) Log.error(error);
                    if (stderr) Log.error(stderr);
                    resolve();
                },
            );

            childProcess.stdout?.pipe(process.stdout);
        });
    }
}
```

**Comment by MarvinRudolph** [comment]
I create models programmatically via files and modify the schema based on that. Then I invoke `migrate` with `execa`.
It would be cool if there would be a function which can be called, something similar to how it works with the generator (I can import and call a function which generates the prisma client).

At the moment, the invocation of the `migrate` command logs messages which is cool if you invoke it by yourself but I want to handle that if I use it programmatically. So no info, success or error messages should appear.
